x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java

```diff
@@ -432,7 +432,18 @@ private Terms wrapTerms(Terms terms, String field) throws IOException {
             // for the _field_names field, fields for the document
             // are encoded as postings, where term is the field.
             // so we hide terms for fields we filter out.
-            return fieldNamesFilterTerms;
+            if (fieldNamesFilterTerms == null) {
+                synchronized (this) {
+                    if (fieldNamesFilterTerms == null) {
+                        assert Transports.assertNotTransportThread("resolving filter terms");
+                        final Terms fieldNameTerms = super.terms(FieldNamesFieldMapper.NAME);
+                        this.fieldNamesFilterTerms = Optional.ofNullable(
+                            fieldNameTerms == null ? null : new FieldNamesTerms(fieldNameTerms)
+                        );
```

Comment:
nit: it feels a bit weird to use `Optional.ofNullable` when you need to check for null already anyway, what about something like that instead?

```java
this.fieldNamesFilterTerms = fieldNameTerms == null ? Optional.empty() : Optional.of(new FieldNamesTerms(fieldNameTerms));
```